mautic/mautic
# Custom field analysis command

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bugfix? | No
| New feature? |  Yes
| Automated tests included? | Yes
| Related user documentation PR URL |
| Related developer documentation PR URL |
| Issues addressed (#s or URLs) |
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The following is the user story this PR addresses.  
_As an_ **engineer**
_I want to_ **Analyse the actual usage of columns (custom fields) in the leads table**
_So that I can_ **set the appropriate size for VARCHAR columns**.

Whenever the user creates any of the fields, the column gets added to the leads table. The default length for a few fields is 64 characters, and it's 255 characters for a few. The following table summarises it,

| Field Type | DB type | Length
|------------| ------- | -----
| Timezone   | VARCHAR | 255
| Locale     | VARCHAR | 255
| Country    | VARCHAR | 255
| Email      | VARCHAR | 64
| Lookup     | VARCHAR | 255
| Select     | VARCHAR | 64
| Region     | VARCHAR | 255
| Phone      | VARCHAR | 255
| Text       | VARCHAR | 64

This command help analyzes all custom fields (type listed in the above table) and suggests the new length.
The command will generate comma-separated values that will help one to store it as a CSV  file. Otherwise, use `-t` to see the output in tabular format.
The headings in the output,
- Custom Field Name: Field label
- Custom Field Alias: Alias of the field
- Current Size: The character length of the fields (varchar length)
- Currently Utilised max size: Max character length of actual field/column record.
- Suggested max size: Minimum of _Currently Utilised max size" * 2_ or _Current Size_ or 191


#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create custom fields
3. Fill in data for custom fields
4. ssh to the webhead,
5. run `bin/console mautic:fields:analyse -t` to see the tabular output
6. Or run `bin/console mautic:fields:analyse > /path/to/csv/file.csv` to save output as CSV file.


[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. Functional test for the command
2. functional test for FieldModel and LeadModel


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10775"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
